File: src/echo-actors/src/main/scala/echo/actor/directory/repository/EpisodeRepository.scala
package echo.actor.directory.repository

import echo.core.domain.entity.{Episode, Podcast}
import org.springframework.data.jpa.repository.{JpaRepository, Query}
import org.springframework.data.repository.query.Param

/**
  * @author Maximilian Irro
  */
trait EpisodeRepository extends JpaRepository[Episode, java.lang.Long] {

    def findOneByEchoId(echoId: String): Episode

    def findAllByPodcast(podcast: Podcast): java.util.List[Episode]

    @Query("SELECT DISTINCT episode FROM Episode episode WHERE episode.podcast.echoId = :podcastExo")
    def findAllByPodcastEchoId(@Param("podcastExo") podcastExo: String): java.util.List[Episode]

    @Query("SELECT DISTINCT episode FROM Episode episode WHERE episode.podcast.echoId = :podcastExo AND episode.guid = :guid")
    def findAllByPodcastAndGuid(@Param("podcastExo") podcastExo: String,
                                @Param("guid") guid: String): java.util.List[Episode]

    @Query("SELECT DISTINCT episode FROM Episode episode " + "" +
           "WHERE episode.enclosureUrl = :enclosureUrl " +
           "AND episode.enclosureLength = :enclosureLength " +
           "AND episode.enclosureType = :enclosureType")
    def findOneByEnlosure(@Param("enclosureUrl") enclosureUrl: String,
                          @Param("enclosureLength") enclosureLength: Long,
                          @Param("enclosureType") enclosureType: String): Episode

    @Query("SELECT count(episode) FROM Episode episode")
    def countAll(): Long

    @Query("SELECT count(episode) FROM Episode episode WHERE episode.podcast.echoId = :podcastExo")
    def countByPodcast(@Param("podcastExo") podcastExo: String): Long
}

File: src/echo-microservices/app-catalog/src/main/java/echo/microservice/catalog/repository/EpisodeRepository.java
package echo.microservice.catalog.repository;

import echo.core.domain.entity.EpisodeEntity;
import echo.core.domain.entity.PodcastEntity;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;

/**
 * @author Maximilian Irro
 */
public interface EpisodeRepository extends JpaRepository<EpisodeEntity,Long> {

    EpisodeEntity findOneByExo(String exo);

    List<EpisodeEntity> findAllByPodcast(PodcastEntity podcast);

    @Query("SELECT DISTINCT episode FROM EpisodeEntity episode WHERE episode.podcast.exo = :exo")
    List<EpisodeEntity> findAllByPodcastExo(@Param("exo") String exo);

    @Query("SELECT DISTINCT episode FROM EpisodeEntity episode WHERE episode.podcast.exo = :podcastExo AND episode.guid = :guid")
    List<EpisodeEntity> findAllByPodcastAndGuid(@Param("podcastExo") String podcastExo,
                                                @Param("guid") String guid);

    @Query("SELECT DISTINCT episode FROM EpisodeEntity episode " + "" +
            "WHERE episode.enclosureUrl = :enclosureUrl " +
            "AND episode.enclosureLength = :enclosureLength " +
            "AND episode.enclosureType = :enclosureType")
    EpisodeEntity findOneByEnlosure(@Param("enclosureUrl") String enclosureUrl,
                                    @Param("enclosureLength") Long enclosureLength,
                                    @Param("enclosureType") String enclosureTypeg);

    @Query("SELECT count(episode) FROM EpisodeEntity episode")
    Long countAll();

    @Query("SELECT count(episode) FROM EpisodeEntity episode WHERE episode.podcast.exo = :podcastExo")
    Long countByPodcast(@Param("podcastExo") String podcastExo);

}
